code/datums/ai/ai.dm

```diff
@@ -1,5 +1,27 @@
 /datum/ai
 	var/name
+	var/mob/living/body		// The parent mob we control.
+	var/wait_for = 	0 		// The next time we can process.
+	var/run_interval = 1 	// How long to wait between processes.
+	
+/datum/ai/New(var/mob/living/body)
+	src.body = body
 
-/datum/ai/proc/process()
-	return PROCESS_KILL
+/datum/ai/proc/can_process()
+	if((body.client || body.mind) && !(body.status_flags & ENABLE_AI))
+		return FALSE
+	if(wait_for > world.time)
+		return FALSE
+	if(body.stat == DEAD)
+		return FALSE
+	wait_for = world.time + run_interval
```

Comment:
A proper method that by its name implies that it only checks something (i.e.: Can Conduct Action, Is Something) should never cause a side-effects (That is: Change variable values, write to file, raise an event, etc.)

This is the expected behavior in general and one should not have to read the code of the proc to avoid confusion/surprise. It also makes it impractical to call such a method multiple times, if ever needed.

Better would be to have the proposed `/datum/ai/proc/process()` handle it via
```
/datum/ai
    var/next_process = 0
    var/run_interval = 1

/datum/ai/proc/can_process()
    if(next_process < world.time)
        return FALSE
    [...]

/datum/ai/proc/process()
    if(can_process())
        do_process()
    next_process = world.time + run_interval
```

I have to wonder why you have the `wait_for`/`run_interval`-vars though. It's already called in intervals via `/ Life()`